idea/frontend/src/features/dns/sails/index.ts

```diff
@@ -0,0 +1,114 @@
+  const { account } = useAccount();
+  const alert = useAlert();
+  const { api } = useApi();
+
+  const sendMessage = async ({ getTransactionBuilder, options }: SendDnsMessage) => {
+    if (!account || !api) {
+      return;
+    }
+    const { resolve: onSuccess, reject: onError } = options || {};
+    const { signer } = await web3FromSource(account.meta.source);
+    // ! TODO: calculate fee by sails
+    // const { partialFee } = await api.message.paymentInfo(account.address, { signer });
+
+    return new Promise<void>((resolve, reject) => {
+      const onConfirm = async () => {
+        if (account) {
+          try {
+            enableLoading();
+            const transaction = await getTransactionBuilder();
+
+            transaction.withAccount(account.address, { signer });
+
+            await transaction.calculateGas();
+            const result = await transaction.signAndSend();
+            await result.response().then(() => {
+              if (onSuccess) onSuccess();
+            });
+            return resolve();
+          } catch (error) {
+            const errorMessage = (error as Error).message;
+            alert.error(errorMessage);
+            if (onError) onError();
+            return reject();
+          } finally {
+            disableLoading();
+          }
+        }
+      };
+
+      showModal('transaction', {
+        fee: '0',
+        name: TransactionName.SendMessage,
+        addressFrom: account.address,
+        addressTo: dnsProgram.programId,
+        onAbort: reject,
+        onConfirm,
+      });
+    });
+  };
+
+  const addNewProgram = (name: string, program_id: HexString, options?: ResolveRejectOptions) => {
+    const getTransactionBuilder = async () => dnsProgram.dns.addNewProgram(name, program_id);
+    return sendMessage({ getTransactionBuilder, options });
+  };
+
+  const changeProgramId = (name: string, program_id: HexString, options?: ResolveRejectOptions) => {
+    const getTransactionBuilder = async () => dnsProgram.dns.changeProgramId(name, program_id);
+    return sendMessage({ getTransactionBuilder, options });
+  };
+
+  const deleteProgram = (name: string, options?: ResolveRejectOptions) => {
+    const getTransactionBuilder = async () => dnsProgram.dns.deleteProgram(name);
+    return sendMessage({ getTransactionBuilder, options });
+  };
+
+  return { isLoading, addNewProgram, changeProgramId, deleteProgram };
+};
+
+const useInitDnsProgram = () => {
+  const { isApiReady, api } = useApi();
+  const [isLoading, setIsLoading] = useState(false);
+
+  useEffect(() => {
+    if (!dnsProgram && !isLoading && isApiReady && api) {
+      setIsLoading(true);
+
+      fetch(`${DNS_API_URL}/dns/contract`).then((response) => {
+        response.json().then(({ contract }) => {
+          const programId = contract;
+          dnsProgram = new Program(api, programId);
+          setIsLoading(false);
+        });
+      });
+    }
+  }, [isApiReady, api]);
+};
```

Comment:
I have a feeling that setting global variable in a `useEffect` is an anti-pattern and we can encounter problems because of it. Let's find out how to deal with it in a react way.

Also, since we are working with async state there, would be better to use `react-query` instead of `useEffect`.